src/Hamiltonians/HOCartesian2BosonRelative.jl

```diff
@@ -0,0 +1,230 @@
+    to be in trap units.
+* `interaction_only`: if set to `true` then the noninteracting single-particle terms are 
+    ignored. Useful if only energy shifts due to interactions are required.
+
+See also [`HOCartesianEnergyConserved`](@ref) and[`HOCartesianEnergyConservedPerDim`](@ref).
+"""
+struct HOCartesian2BosonRelative{
+    D,  # number of dimensions
+    A<:BoseFS
+} <: AbstractHamiltonian{Float64}
+    addr::A
+    S::NTuple{D,Int64}
+    aspect::NTuple{D,Float64}
+    vtable::Vector{Float64}  # interaction coefficients
+    g::Float64
+    interaction_only::Bool
+end
+
+function HOCartesian2BosonRelative(
+        addr; # ignored for this Hamiltonian if no other arguments are provided
+        S = (num_modes(addr),),
+        Sx = nothing,
+        ηs = nothing, 
+        g = 1.0,
+        interaction_only = false,
+        warn_address = true
+    )
+    if isnothing(ηs) && isnothing(Sx)
+        S[1] ≠ maximum(S) && throw(ArgumentError("First dimension must have the most states"))
+        aspect = float.(box_to_aspect(S))
+        D = length(S)
+    else
+        any(ηs .< 1.0) && throw(ArgumentError("Aspect ratios must be greater than 1.0"))
+        D = length(ηs) + 1
+        Srest = map(x -> floor(Int, (Sx - 1) / x + 1), ηs)
+        S = (Sx, Srest...)
+        aspect = (1.0, float.(ηs)...)
+    end
+    P = prod(S)
+
+    # address type is fixed so it is overwritten here to ensure it works
+    warn_address && !(addr isa BoseFS{1,P}) && @warn "Input address is being overwritten"
+    addr = BoseFS(P, 1 => 1)
+
+    M = maximum(S) - 1
+    v_vec = [log_abs_oscillator_zero(i) for i in 0:2:M]
+
+    return HOCartesian2BosonRelative{D,typeof(addr)}(addr, S, aspect, v_vec, g, interaction_only)
+end
+
+function Base.show(io::IO, H::HOCartesian2BosonRelative)
+    if length(H.S) == 1
+        print(io, "HOCartesian2BosonRelative($(H.addr); S=$(H.S), g=$(H.g), interaction_only=$(H.interaction_only))")
+    else
+        print(io, "HOCartesian2BosonRelative($(H.addr); Sx=$(H.S[1]), ηs=$(H.aspect[2:end]), g=$(H.g), interaction_only=$(H.interaction_only))")
+    end
+end
+
+Base.:(==)(H::HOCartesian2BosonRelative, G::HOCartesian2BosonRelative) = all(map(p -> getproperty(H, p) == getproperty(G, p), propertynames(H)))
+
+starting_address(H::HOCartesian2BosonRelative) = H.addr
+
+LOStructure(::Type{<:HOCartesian2BosonRelative}) = IsHermitian()
+
+### DIAGONAL ELEMENTS ###
+@inline function noninteracting_energy(S, aspect, omm::BoseOccupiedModeMap)
+    states = CartesianIndices(S)
+    return sum(omm) do p
+        indices = Tuple(states[p.mode]) .- 1
+        dot(indices, aspect) + sum(aspect)/2
+    end
+end
+noninteracting_energy(H::HOCartesian2BosonRelative, addr) = noninteracting_energy(H.S, H.aspect, OccupiedModeMap(addr))
+
+@inline function diagonal_element(H::HOCartesian2BosonRelative, addr)
+    omm = OccupiedModeMap(addr)
+    u = H.g / sqrt(prod(H.aspect))
+    result = u * ho_2brel_interaction(H.S, H.vtable, omm, omm)
```

Comment:
If you change `ho_2brel_interaction` to accept orbital indices (instead of `omm`s) as suggested in my other comment, then this line needs to be changed into a sum over all orbitals in the `omm`.  Something like
```suggestion
    result = u * sum(omm) do p
        p.occnum * ho_2brel_interaction(H.S, H.vtable, p, p)
    end
```